src/backend/distributed/planner/query_colocation_checker.c

```diff
@@ -361,8 +362,9 @@ CreateAllTargetListForRelation(Oid relationId, List *requiredAttributes)
 		}
 		else
 		{
+			int varAttNum = isMergeQuery ? attrNum : colAppendIdx++;
```

Comment:
if you really meant to use the exact column position whether selected or not, I think you are looking at `int resNo = attrNum;` so you can use this for merge